### ISSUE TITLE
`ByteStreams.nullByteSink()`

### ISSUE THREAD

**Issue by fror** [google/guava#2011]
There exists a null `OutputStream` with `ByteStreams.nullOutputStream()`. It would be nice to have the counterpart `ByteStreams.nullByteSink()` that has methods that would specifically not do anything.

The use case I have is the following:

I have files for which I want to compute the `HashCode` and only that. So I'm using this code:

    try (HashingInputStream in = new HashingInputStream(Hashing.sha1(), Files.newInputStream(path))) {
      while (in.read() >= 0) { /* do nothing */ }
      return in.hash();
    }

Ideally I want to read those bytes but not doing anything. This code is great, does it as expected, but it is not really understandable at first glance that I'm actually skipping all of those. I checked some alternatives, like `ByteStreams.copy(in, ByteStreams.nullOutputStream())` or `ByteStreams.read(in, NullByteProcessor.INSTANCE)`, with `NullByteProcessor` doing what you think it does, but those always call several methods. So, how to wrap that null read properly?

I saw that `ByteSink.writeFrom(InputStream)` exists and I thought that maybe a `NullByteSink.writeFrom(InputStream)` is the exact proper way to wrap the `while (in.read() >= 0) { /* do nothing */ }` loop.

If it were implemented, I don't really know where I'd put that NullByteSink: in `ByteStreams.nullByteSink()`? In `ByteSink.nullByteSink()`? Your call ;)

**Comment by kluever** [comment]
Have you seen `Files.hash(File, HashFunction)`?

**Comment by fror** [comment]
No, I had not seen that. And it will be absolutely perfect for me, thanks!

However, that's a bit something I find awful with Guava. All the functionalities are available, but are nowhere to be found. For instance, I'm using paths using Java 7-8 nio, so I'm naturally drawn to use something common between those API and what I know of Guava, the streams. Once I have the streams, well I couldn't find any solution. I had checked both [HashingExplained](https://code.google.com/p/guava-libraries/wiki/HashingExplained) and the [`c.g.c.hash package`](http://docs.guava-libraries.googlecode.com/git-history/release/javadoc/com/google/common/hash/package-summary.html) and none of those ever mentioned [`Files.hash(File, HashFunction)`](http://docs.guava-libraries.googlecode.com/git-history/release/javadoc/com/google/common/io/Files.html#hash(java.io.File,%20com.google.common.hash.HashFunction)) Of course, going the good old "File" way solves this, but I kind of try avoid using files ever since I use Java 7, like a lot of people, to be honest, as I can see on StackOverflow.

I have the answer to my use case yet I leave the ticket open, so if you think about nice other use cases of the null `ByteSink`, just keep it open.

**Comment by kluever** [comment]
Sorry about the lack of documentation. It's tough for us to mention every single API in the "Guava Explained" wiki.

@cgdecker Any thoughts on the idea of a nullByteSink?

**Comment by cgdecker** [comment]
@kluever I believe I proposed a `nullByteSink` at some point but we decided not to do it unless there was some need for it.

With regard to @fror's use case: I think that clearly, the best thing here* would be for us to release `MoreFiles`, which would allow `MoreFiles.asByteSource(path).hash(Hashing.sha1())`. Barring that, using `ByteStreams.copy(in, ByteStreams.nullOutputStream())` doesn't seem that bad. There's also a question of whether we should have something like `ByteStreams.hash(InputStream, HashFunction)`.

* I still think that ideally `common.hash` should depend on `common.io` rather than the other way around and that `Hasher` and `HashFunction` should have methods for hashing an `InputStream`, `ByteSource`, etc. Though that's likely impossible to change at this point.

**Comment by fror** [comment]
@kluever No need to apologize. Guava is extremely well done and thought. Indeed the documentation is sometimes a bit lacking, but there's nothing to apologize for. ;)

**Comment by Marcono1234** [comment]
A little bit similar to what I requested in #3522. It appears currently a good solution is to use `ByteStreams.exhaust(InputStream)`, which likely performs better than calling `read()` in a loop, since that only reads one byte at a time.
